### PR TITLE
Introduce a configuration framework for better organizing and documentating configuration settings

### PR DESCRIPTION
**Pull request by sijie**
Descriptions of the changes in this PR:

### Motivation

One common task in developing bookkeeper is to make sure all the configuration
settings are well documented, and the configuration file we ship in each release
is in-sync with the code itself.

However maintaining things in-sync is non-trivial. This proposal is exploring
a new way to manage configuration settings for better documentation.

### Changes

- Introduce `ConfigKey` for defining a configuration setting key in a configuration
- Introduce `ConfigKeyGroup` for grouping configuration settings together
- Introduce `ConfigDef` for generating the configuration definition for a given configuration
- Add a `save` method for saving a configuration definition into a configuration file

Master Issue: #1867 

> ---
> In order to uphold a high standard for quality for code contributions, Apache BookKeeper runs various precommit
> checks for pull requests. A pull request can only be merged when it passes precommit checks. However running all
> the precommit checks can take a long time, some trivial changes don't need to run all the precommit checks. You
> can check following list to skip the tests that don't need to run for your pull request. Leave them unchecked if
> you are not sure, committers will help you:
>
> - [ ] [skip bookkeeper-server bookie tests]: skip testing `org.apache.bookkeeper.bookie` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server client tests]: skip testing `org.apache.bookkeeper.client` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server replication tests]: skip testing `org.apache.bookkeeper.replication` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server tls tests]: skip testing `org.apache.bookkeeper.tls` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server remaining tests]: skip testing all other tests in bookkeeper-server module.
> - [ ] [skip integration tests]: skip docker based integration tests. if you make java code changes, you shouldn't skip integration tests.
> - [ ] [skip build java8]: skip build on java8. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> - [ ] [skip build java9]: skip build on java9. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> ---

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue #>: Description of pull request`
>     `e.g. Issue 123: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue #>` in the title with the actual Issue number.
> 
> ---

### CODE REVIEW

**Comment by sijie** [comment]
@merlimat : since bk's configuration is using commons configuration, it is quite different from pulsar's configuration. so I didn't use pulsar's annotation approach here, because I feel if we want to use annotations, we have to bk's configration, which can be a much bigger change. so I keep the current configuration's format, but change how it defines configuration keys and groups.

File: bookkeeper-common/src/main/java/org/apache/bookkeeper/common/conf/ConfigDef.java [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+    public void save(Path path) throws IOException  {
+        try (OutputStream stream = Files.newOutputStream(
+            path, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING)) {
+            save(stream);
+        }
+    }
+
+    public void save(OutputStream os) {
+        try (PrintStream ps = new PrintStream(os)) {
```
Comment (eolivelli) [inline bookkeeper-common/src/main/java/org/apache/bookkeeper/common/conf/ConfigDef.java]: Is this Platform dependant? Thinking about EOL

Comment (sijie) [reply bookkeeper-common/src/main/java/org/apache/bookkeeper/common/conf/ConfigDef.java]: EOL should be already handled. what is your concern here?

Comment (eolivelli) [reply bookkeeper-common/src/main/java/org/apache/bookkeeper/common/conf/ConfigDef.java]: If you are saving this file on linux you wil use LF, on Mac CR...
Maybe this is not a real problem, as the original file will already be using such EOL convention

Comment (sijie) [reply bookkeeper-common/src/main/java/org/apache/bookkeeper/common/conf/ConfigDef.java]: that has to be platform dependent. so I don't think that is an issue.

**Comment by sijie** [comment]
run bookkeeper-server remaining tests
run integration tests

**Comment by sijie** [comment]
run integration tests

**Comment by sijie** [comment]
rebuild java9
run integration tests

**Pull request merged**